onnxruntime/core/providers/cpu/nn/batch_norm.h

```diff
@@ -28,28 +28,40 @@
 
 namespace onnxruntime {
 
+#if !defined(ORT_MINIMAL_BUILD)
+#define BATCHNORM_INCLUDE_TRAINING_SUPPORT
+#endif
+
 template <typename T>
 class BatchNorm : public OpKernel {
  public:
   explicit BatchNorm(const OpKernelInfo& op_kernel_info)
-      : OpKernel(op_kernel_info), is_spatial_(op_kernel_info.GetAttrOrDefault<int64_t>("spatial", 1) == 1) {
-    epsilon_ = op_kernel_info.GetAttrOrDefault<float>("epsilon", 1e-5f);
-    momentum_ = op_kernel_info.GetAttrOrDefault<float>("momentum", 0.9f);
-
+      : OpKernel(op_kernel_info),
+        epsilon_(op_kernel_info.GetAttrOrDefault<float>("epsilon", 1e-5f)),
+        is_spatial_(op_kernel_info.GetAttrOrDefault<int64_t>("spatial", 1) == 1) {
     // For opset 6-8, if spatial attribute exists, pick up the value (by default spatial == 1)
     // From opset 9 onwards, by default, only the spatial case (spatial == 1) is defined per spec
     // For opset 14 onwards, training is an attribute.
     // For opset < 14, since no training attribute is present we assume optional outputs indicate training mode.
     if (op_kernel_info.node().SinceVersion() == 14) {
       is_train_ = op_kernel_info.GetAttrOrDefault<int64_t>("training_mode", 0) == 1;
-      size_t output_count = op_kernel_info.node().OutputDefs().size();
-      ORT_ENFORCE((is_train_ && output_count == 3) || (!is_train_ && output_count == 1),
-                  "Output running_mean and running_var are valid and required for training mode.");
     } else {
       is_train_ = OpKernel::Node().OutputDefs().size() > 1;
     }
 
-    ORT_ENFORCE(!is_train_ || is_spatial_, "Training mode does not support non-spatial BN");
+    if (is_train_) {
+#if defined(BATCHNORM_INCLUDE_TRAINING_SUPPORT)
+      size_t output_count = op_kernel_info.node().OutputDefs().size();
+      ORT_ENFORCE((is_train_ && output_count == 3) || (!is_train_ && output_count == 1),
```

Comment:
Move this outside of ` if (is_train_)`?
And isn't `output_count == 3` opset 14 only?

Reply:
You're right - I overlooked it changing from 5 to 3. The ONNX type/shape inferencing checks the number of outputs is valid for opset 14 though so we don't need to do that again here.

Based on that I don't think the ORT_ENFORCE is needed as we didn't have it prior to opset 14.